python/cudf/cudf/pylibcudf_tests/common/utils.py

```diff
@@ -102,14 +104,56 @@ def _make_fields_nullable(typ):
             return pa.list_(new_fields[0])
         return typ
 
+    def _contains_type(parent_typ, typ_checker):
+        """
+        Check whether the parent or one of the children
+        satisfies the typ_checker.
+        """
+        if typ_checker(parent_typ):
+            return True
+        if pa.types.is_nested(parent_typ):
+            for i in range(parent_typ.num_fields):
+                if _contains_type(parent_typ.field(i).type, typ_checker):
+                    return True
+        return False
+
     if not check_field_nullability:
         rhs_type = _make_fields_nullable(rhs.type)
         rhs = rhs.cast(rhs_type)
 
         lhs_type = _make_fields_nullable(lhs.type)
-        lhs = rhs.cast(lhs_type)
+        lhs = lhs.cast(lhs_type)
 
-    assert lhs.equals(rhs)
+    assert lhs.type == rhs.type, f"{lhs.type} != {rhs.type}"
```

Comment:
This whole thing is part of assert_column_eq.

This particular assert is to provide more debug information when the array types don't match.

The assert array.equals will only tell you below that arrays don't match (which either could be because values don't match or because the type doesn't match)